stan/math/prim/fun/reverse.hpp

```diff
@@ -0,0 +1,53 @@
+#ifndef STAN_MATH_PRIM_FUN_REVERSE_HPP
+#define STAN_MATH_PRIM_FUN_REVERSE_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <algorithm>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return a copy of the specified array in reversed order.
+ *
+ * @tparam T type of elements in the array
+ * @param x array to reverse
+ * @return Array in reversed order.
+ */
+template <typename T>
+inline std::vector<T> reverse(const std::vector<T>& x) {
+  std::vector<T> rev(x.size());
+  std::reverse_copy(x.begin(), x.end(), rev.begin());
+  return rev;
+}
+
```

Comment:
You could replace the three definitions with a single ```apply_vector_unary``` here:
```
template <typename T>
inline auto reverse(const T& x) {
  return apply_vector_unary<T>::apply(x, [&](const auto& v) {
    return v.reverse();
  });
}
````

Reply:
I should note that this would also implicitly allow for ```reverse``` with nested arrays (e.g. ```reverse(std::vector<VectorXd>)```) which returns an array with each vector reversed

Reply:
Good to know! But for `std::vector`, will this make a copy?

Reply:
It has the same cost as the current implementations, where the result is evaluated into a vector which is then returned. ```Eigen::Map``` is used to evaluate directly into an ```std::vector``` and avoid the need for an intermediary copy

Reply:
> I should note that this would also implicitly allow for reverse with nested arrays

I think this would be super confusing.  I think there should be a reverse function that works the same for any type `T[]`, namely by reversing the elements.  If the `T` is a container type, I don't think it should switch to working elementwise.

If there really needs to be a form of `reverse` that reverses containers within a list, I think it should get a different name.

I feel very strongly that this pattern of behavior should hold everywhere.

Reply:
Sounds good to me, @mcol can you revert this change (sorry for the detour!). You could instead collapse the two Eigen definitions, so that you just have one definition for Eigen vectors (row or col) and one for std::vectors:
```
template <typename T>
inline std::vector<T> reverse(const std::vector<T>& x) {
  std::vector<T> rev(x.size());
  std::reverse_copy(x.begin(), x.end(), rev.begin());
  return rev;
}

template <typename T, require_eigen_vector_t<T>>
inline auto reverse(const T& x) {
  return x.reverse();
}
```

Reply:
No worries, if anything I learned about `apply_vector_unary`! :)

Reply:
Sorry to be a pain, but that eigen definition should have had an ```.eval()``` at the end, otherwise it could cause the same issues as #1644:

```
template <typename T, require_eigen_vector_t<T>>
inline auto reverse(const T& x) {
  return x.reverse().eval();
}
```

Reply:
Thanks!  After posting, I wanted to make sure I wasn't around the bend on this one, so I took an informal poll of the non-computer scientists in my office (I won't name names) and they expected a top-level reverse only, not an elementwise one.